drivers/input/touchscreen/zinitix.c

```
@@ -0,0 +1,670 @@
+	}
+	if (zinitix_read_data(client, BT541_THRESHOLD, (u8 *)&cap->threshold, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_THRESHOLD,
+					(u8 *)&cap->threshold, 2) < 0)
+			return 1;
+
+	if (zinitix_read_data(client, BT541_BUTTON_SENSITIVITY,
+					(u8 *)&cap->key_threshold, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_DUMMY_BUTTON_SENSITIVITY,
+					(u8 *)&cap->dummy_threshold, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_TOTAL_NUMBER_OF_X,
+					(u8 *)&cap->x_node_num, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_TOTAL_NUMBER_OF_Y,
+					(u8 *)&cap->y_node_num, 2) < 0)
+		return 1;
+
+	cap->total_node_num = cap->x_node_num * cap->y_node_num;
+
+	if (zinitix_read_data(client, BT541_DND_N_COUNT,
+					(u8 *)&cap->N_cnt, 2) < 0)
+		return 1;
+
+	printk(KERN_INFO "N count = %d\n", cap->N_cnt);
+
+	if (zinitix_read_data(client, BT541_DND_U_COUNT,
+					(u8 *)&cap->u_cnt, 2) < 0)
+		return 1;
+
+	printk(KERN_INFO "u count = %d\n", cap->u_cnt);
+
+	if (zinitix_read_data(client, BT541_AFE_FREQUENCY,
+					(u8 *)&cap->afe_frequency, 2) < 0)
+		return 1;
+
+	printk(KERN_INFO "afe frequency = %d\n", cap->afe_frequency);
+
+
+	/* get chip firmware version */
+	if (zinitix_read_data(client, BT541_FIRMWARE_VERSION,
+		(u8 *)&cap->fw_version, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_MINOR_FW_VERSION,
+		(u8 *)&cap->fw_minor_version, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_DATA_VERSION_REG,
+		(u8 *)&cap->reg_data_version, 2) < 0)
+		return 1;
+
+	if (zinitix_read_data(client, BT541_EEPROM_INFO_REG,
+			(u8 *)&chip_eeprom_info, 2) < 0)
+		return 1;
+
+	/* initialize */
+	if (zinitix_write_u16(client, BT541_X_RESOLUTION,
+			(u16)540/*pdata->x_resolution*/) < 0)
+		return 1;
+
+	if (zinitix_write_u16(client, BT541_Y_RESOLUTION,
+			(u16)960/*pdata->y_resolution*/) < 0)
+		return 1;
+
+	cap->MinX = (u32)0;
+	cap->MinY = (u32)0;
+	cap->MaxX = (u32)540/*pdata->x_resolution*/;
```

Avoid hardcoding

---

I'm using that, but it was run after that function and now I will probably have to remove the input device on error?

---

`touchscreen_parse_properties` does not return an error, although you should check that the coordinates are set after running it.

Since you use `devm_input_allocate_device`, you only need to return an error from the probe method and the input device will be removed automagically.

---

touchscreen_parse_properties seems to want the input dev pointer to be non-NULL

---

Correct.